server/src/main/java/org/opensearch/indices/IndicesRequestCache.java

```diff
@@ -205,6 +235,68 @@ public void onHit(Key key, CacheValue<BytesReference> cacheValue) {
     @Override
     public void onCached(Key key, BytesReference value, TierType tierType) {
         key.entity.onCached(key, value, tierType);
+        updateDiskCleanupKeyToCountMap(new CleanupKey(key.entity, key.readerCacheKeyId), tierType);
+    }
+
+    /**
+     * Updates the diskCleanupKeyToCountMap with the given CleanupKey and TierType.
+     * If the TierType is not DISK, the method returns without making any changes.
+     * If the ShardId associated with the CleanupKey does not exist in the map, a new entry is created.
+     * The method increments the count of the CleanupKey in the map.
+     *
+     * Why use ShardID as the key ?
+     * CacheEntity mainly contains IndexShard, both of these classes do not override equals() and hashCode() methods.
+     * ShardID class properly overrides equals() and hashCode() methods.
+     * Therefore, to avoid modifying CacheEntity and IndexShard classes to override these methods, we use ShardID as the key.
+     *
+     * @param cleanupKey the CleanupKey to be updated in the map
+     * @param tierType the TierType of the CleanupKey
+     */
+    private void updateDiskCleanupKeyToCountMap(CleanupKey cleanupKey, TierType tierType) {
+        if(!tierType.equals(TierType.DISK)) {
+            return;
+        }
+        IndexShard indexShard = (IndexShard)cleanupKey.entity.getCacheIdentity();
+        ShardId shardId = indexShard.shardId();
+
+        diskCleanupKeyToCountMap
+            .computeIfAbsent(shardId, k -> ConcurrentCollections.newConcurrentMap())
+            .merge(cleanupKey.readerCacheKeyId, 1, Integer::sum);
+    }
+
+    /**
+     * Updates the count of stale keys in the disk cache.
+     * This method is called when a CleanupKey is added to the keysToClean set.
+     * It increments the staleKeysInDiskCount by the count of the CleanupKey in the diskCleanupKeyToCountMap.
+     * If the CleanupKey's readerCacheKeyId is null or the CleanupKey's entity is not open, it increments the staleKeysInDiskCount
+     * by the total count of keys associated with the CleanupKey's ShardId in the diskCleanupKeyToCountMap and removes the ShardId from the map.
+     *
+     * @param cleanupKey the CleanupKey that has been marked for cleanup
+     */
+    private void updateStaleKeysInDiskCount(CleanupKey cleanupKey) {
+        IndexShard indexShard = (IndexShard) cleanupKey.entity.getCacheIdentity();
+        ShardId shardId = indexShard.shardId();
+
+        ConcurrentMap<String, Integer> countMap = diskCleanupKeyToCountMap.get(shardId);
+        if (countMap == null) {
+            return;
+        }
+
+        if (cleanupKey.readerCacheKeyId == null) {
```

Comment:
When will the reader cache key ID be null? Is this when the shard is totally closed?